omnigibson/scene_graphs/graph_builder.py

```diff
@@ -0,0 +1,272 @@
+                    # TODO: Maybe the states actually need to be robust to bad arguments.
+                    pass
+
+    return states
+
+
+class SceneGraphBuilder(object):
+    def __init__(self, egocentric=False, full_obs=False, only_true=False, merge_parallel_edges=False):
+        """
+        @param egocentric: Whether the objects should have poses in the world frame or robot frame.
+        @param full_obs: Whether all objects should be updated or only those in FOV of the robot.
+        @param only_true: Whether edges should be created only for relative states that have a value True, or for all
+            relative states (with the appropriate value attached as an attribute).
+        @param merge_parallel_edges: Whether parallel edges (e.g. different states of the same pair of objects) should
+            exist (making the graph a MultiDiGraph) or should be merged into a single edge instead.
+        """
+        self.G = None
+        self.egocentric = egocentric
+        self.full_obs = full_obs
+        self.only_true = only_true
+        self.merge_parallel_edges = merge_parallel_edges
+        self.last_desired_frame_to_world = None
+
+    def get_scene_graph(self):
+        return self.G.copy()
+
+    def _get_desired_frame(self, robot):
+        desired_frame_to_world = np.eye(4)
+        world_to_desired_frame = np.eye(4)
+        if self.egocentric:
+            desired_frame_to_world = get_robot_to_world_transform(robot)
+            world_to_desired_frame = T.pose_inv(desired_frame_to_world)
+
+        return desired_frame_to_world, world_to_desired_frame
+
+    def start(self, scene):
+        assert self.G is None, "Cannot start graph builder multiple times."
+
+        robot = get_robot(scene)
+        self.G = nx.DiGraph() if self.merge_parallel_edges else nx.MultiDiGraph()
+
+        desired_frame_to_world, world_to_desired_frame = self._get_desired_frame(robot)
+        robot_pose = world_to_desired_frame @ get_robot_to_world_transform(robot)
+        robot_bbox_pose, robot_bbox_extent = get_robot_bbox(robot)
+        robot_bbox_pose = world_to_desired_frame @ robot_bbox_pose
+        self.G.add_node(
+            robot, pose=robot_pose, bbox_pose=robot_bbox_pose, bbox_extent=robot_bbox_extent, states={}
+        )
+        self.last_desired_frame_to_world = desired_frame_to_world
+
+    def step(self, scene):
+        assert self.G is not None, "Cannot step graph builder before starting it."
+
+        # Prepare the necessary transformations.
+        robot = get_robot(scene)
+        desired_frame_to_world, world_to_desired_frame = self._get_desired_frame(robot)
+
+        # Update the position of everything that's already in the scene by using our relative position to last frame.
+        old_desired_to_new_desired = world_to_desired_frame @ self.last_desired_frame_to_world
+        for obj in self.G.nodes:
+            self.G.nodes[obj]["pose"] = old_desired_to_new_desired @ self.G.nodes[obj]["pose"]
```

Comment:
nit: do we need this info? if we just have the bbox_pose isn't that sufficient to back out the corresponding pose if necessary? this would save us half the number of TFs.

It would be even better if we do a batched operation -- i.e.: tensorized matrix multiplication so we only do exactly one computation per step. You can see what's done in the macro_particle_system, and it saves a significant amount of time once the quantity becomes large (even 10s if IIRC)

Reply:
I think we should keep the additional info - it's not clear in the non-OG context. Agreed re: vectorization.